pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -239,77 +241,72 @@ func (d *GPUDcgm) GetProcessResourceUtilizationPerDevice(device interface{}, gpu
-						Pid:         p.Pid,
-						TimeStamp:   uint64(time.Now().UnixNano()),
-						ComputeUtil: uint32(normalizedComputeUtil),
+		isMIG := p.GpuInstanceId != 0 && p.GpuInstanceId < maxMIGProfiles // if this is not a MIG, GpuInstanceId is set to e.g. 4294967295
+		if isMIG {                                                        // this is a MIG, get it entity id and reads the related fields
+			// get the MIG info from d.entities
+			for _, migInfo := range d.entities[gpuID] {
+				if p.GpuInstanceId == uint32(migInfo.Info.NvmlInstanceId) {
+					klog.V(debugLevel).Infof("MIG Device %v InstanceId %v GpuIndex %v \n",
+						migInfo.Entity.EntityId, migInfo.Info.NvmlInstanceId, migInfo.Info.NvmlGpuIndex)
+					vals, err = dcgm.EntityGetLatestValues(dcgm.FE_GPU_I, migInfo.Entity.EntityId, deviceFields)
+					if err != nil {
+						klog.V(debugLevel).Infof("failed to get latest values for fields: %v", err)
+						return processAcceleratorMetrics, err
 					}
+					multiprocessorCountRatio = float64(migInfo.Profile.MultiprocessorCount) / migInfo.TotalMultiprocessorCount
+					break
 				}
 			}
-			klog.V(debugLevel).Infof("\n")
-		} else {
-			gpuUtilization := float64(0)
-			if err == nil {
-				for _, val := range vals {
-					if val.FieldId == ratioField {
-						gpuUtilization = ToFloat64(val, 100)
-					}
+		} else { // this is a GPU, get the GPU utilization
+			multiprocessorCountRatio = 1.0
+			vals, err = dcgm.GetLatestValuesForFields(uint(gpuID), deviceFields)
+		}
+		for _, val := range vals {
+			if val.FieldId == ratioField {
+				deviceUtilization := ToFloat64(val, 100)
+				if deviceUtilization > 100 {
+					deviceUtilization = 100
 				}
+				// ratio of active multiprocessors to total multiprocessors
+				// the MIG metrics represent the utilization of the MIG device. We need to normalize the metric to represent the overall GPU utilization
+				// FIXME: the MIG device could have multiple processes, how to split the MIG utilization between the processes?
+				normalizedComputeUtil := deviceUtilization * multiprocessorCountRatio
+				totalNormalizedUtil += normalizedComputeUtil
+				processNormalizedUtil[p.Pid] = normalizedComputeUtil
+				klog.V(debugLevel).Infof("pid %d computeUtil %f multiprocessor count ratio %v\n", p.Pid, normalizedComputeUtil, multiprocessorCountRatio)
 			}
-			processAcceleratorMetrics[p.Pid] = ProcessUtilizationSample{
-				Pid:       p.Pid,
-				TimeStamp: uint64(time.Now().UnixNano()),
-				// TODO: It does not make sense to use the whole GPU utilization since a GPU might have more than one PID
-				// FIXME: As in the original NVML code, we should use here the pinfo.SmUtil from GetProcessUtilization()
-				ComputeUtil: uint32(gpuUtilization),
-			}
 		}
 	}
+	// fourth, get the process utilization
+	for _, p := range processInfo {
+		computeUtilization := 0.0
+		if totalNormalizedUtil > 0 {
+			computeUtilization = processNormalizedUtil[p.Pid]/totalNormalizedUtil*100 + 0.5
```

Comment:
We should not calculate the Ratio here, but in the power model.
Also, we should not sum the `+ 0.5`

Reply:
Also, this is not really the Ratio.
The Ratio is the sum of the utilization of all MIG slices, but here we have the utilization of only one slice.

Reply:
This function is listing all process running in the MIG slice, and then assuming that there is only one process running in the MIG slice, it associate the MIG utilization to the process.
The MIG utilization is normalized by its size to represent a fraction of the total GPU utilization.


So, for example, if there is only one process using in a MIG slices.
The `totalNormalizedUtil` = `processNormalizedUtil`  and the ratio will be `100.5`

It means that each MIG ratio will 1, and when multiplied by the GPU power it will have the total GPU power, not the fraction of the GPU power.